File: 백준/Gold/14502. 연구소/연구소.cc
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <string>
#include <memory.h>
#include <algorithm>
#include <vector>
#include <stack>
#include <queue>

using namespace std;
#define INF 2147483648
typedef long long ll;

int N, M, ans = 0;
int arr[9][9], tmp[9][9], virus[9][9];
int dx[] = {0, 0, -1, 1}, dy[] = {1, -1, 0, 0};
bool visited[9][9];
queue<pair<int, int>> q;

int bfs(void)
{
    int cnt = 0;
    while (!q.empty())
    {
        int nowX = q.front().first;
        int nowY = q.front().second;
        q.pop();
        visited[nowX][nowY] = true;
        for (int i = 0; i < 4; i++)
        {
            int nextX = nowX + dx[i];
            int nextY = nowY + dy[i];
            if (0 <= nextX && nextX < N && 0 <= nextY && nextY < M && !visited[nextX][nextY])
            {
                if (virus[nextX][nextY] == 0)
                {
                    virus[nextX][nextY] = 2;
                    q.push({nextX, nextY});
                }
            }
        }
    }
    // 안전구역 크기 계산
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            if (virus[i][j] == 0)
                cnt++;
        }
    }
    return cnt;
}

void init(void)
{
    // 초기화 및 벽을 세운 정보 복사하기
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            visited[i][j] = false;
            virus[i][j] = tmp[i][j];
            // 바이러스의 위치 저장
            if (tmp[i][j] == 2)
                q.push({i, j});
        }
    }
}

void makeWall(int idx)
{
    // 임의의 벽 3개 세우기
    if (idx == 3)
    {
        init();
        ans = max(ans, bfs());
        return;
    }
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            if (!tmp[i][j])
            {
                tmp[i][j] = 1;
                makeWall(idx + 1);
                tmp[i][j] = 0;
            }
        }
    }
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    cin >> N >> M;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            cin >> arr[i][j];
            tmp[i][j] = arr[i][j];
        }
    }
    makeWall(0);
    cout << ans << "\n";
    return 0;
}

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static StringBuilder sb = new StringBuilder();
    private static StringTokenizer st;
    private static int N, M, ans = 0;
    private static int[][] arr;
    private static boolean[][] wall;
    private static boolean[][] visited;
    private static int[] dx = {0, 0, -1, 1};
    private static int[] dy = {1, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N + 1][M + 1];
        wall = new boolean[N + 1][M + 1];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        makeWall(0);
        System.out.println(ans);
    }

    private static void makeWall(int cnt) {
        /* 벽을 모두 세운 경우 */
        if (cnt == 3) {
            int[][] tmp = new int[N + 1][];
            for (int i = 0; i < N; i++) {
                tmp[i] = Arrays.copyOf(arr[i], arr[i].length);
            }
            /* bfs */
            bfs(tmp);
            int safeCnt = 0;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (tmp[i][j] == 0) {
                        safeCnt++;
                    }
                }
            }
            ans = Math.max(ans, safeCnt);
            return;
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j] == 0 && !wall[i][j]) {
                    wall[i][j] = true;
                    arr[i][j] = 1;
                    makeWall(cnt + 1);
                    wall[i][j] = false;
                    arr[i][j] = 0;
                }
            }
        }
    }

    private static void bfs(int[][] tmp) {
        Queue<Node> q = new ArrayDeque<>();
        visited = new boolean[N + 1][M + 1];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (tmp[i][j] == 2) {
                    q.offer(new Node(i, j));
                }
            }
        }
        while (!q.isEmpty()) {
            Node node = q.poll();
            int nowX = node.x;
            int nowY = node.y;
            visited[nowX][nowY] = true;
            for (int i = 0; i < 4; i++) {
                int nextX = nowX + dx[i];
                int nextY = nowY + dy[i];
                if (isValid(nextX, nextY) && tmp[nextX][nextY] == 0) {
                    tmp[nextX][nextY] = 2;
                    q.offer(new Node(nextX, nextY));
                }
            }
        }
    }

    private static boolean isValid(int x, int y) {
        return 0 <= x && x < N && 0 <= y && y < M && !visited[x][y];
    }

    static class Node {
        int x;
        int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}